### PR TITLE
Work towards 1.0. Refactoring of runners/events/web ui. Getting rid of global state.

### PR DESCRIPTION
**Pull request by heyman**
Hey!

I've done some refactoring of Locust. The main purpose is to improve the code quality, and make some parts of Locust that is currently non public part of the public API.

It's now in a working state but since it's a few changes to the public API, as well as a lot of changes to the internal API, I think it would be a good idea to have these changes in a 1.0. Therefore - if others doesn't think it's a terrible idea - I suggest that we call a feature freeze for master (and implement all new features in the v1.0 branch) since it'll be a messy job to port new features from master.

Please note that it's a work in progress. All tests pass, but there are no documentation changes yet. I wanted to get it out in a Github PR as soon as possible in order to get more eyes on it. 

### Implemented changes

These are the changes that are currently implemented in this branch:

* An `Environment` class has been introduced to work as a glue between the runner, locust classes, events and web UI. The main benefit of this is that we can get rid of the global `locust.stats.global_stats` and `locust.runner.locust_runner` instances. End users will also be able to run Locust as a lib/programatically (#222), and it'll be easier to write (unit) tests for Locust.
* `locust.events` has been moved to the `events` attribute of the `Environment` instance.
* All `Locust` classes must now be passed an `Environment` instance reference (when instantiated) which is then stored on `Locust.environment`. It's then used by the HTTP client when it triggers `request_success`/`request_failure` events.
* A `WebUI` class has been added in order to be able to spawn one (or multiple) web UI server(s) when running Locust programatically.
* `LocustRunner.start_hatching()` has been renamed to `LocustRunner.start()`
* `LocustRunner.hatch_rate` has been removed in favour of it just being an argument to `start()` and `spawn_locusts()`.

### More changes

Here are other changes that I think we should implement/consider for a 1.0. Most of them have been suggested/requested by others before at some point:

- [x] We need ~~to figure out~~ an API for some initialization hook in end users' test scripts. Since the the test script is imported before the `Environment` is constructed, we can't rely on letting end users put the code at the module level.
  API:
  ```python
  # in end-user's test script
  from locust.events import init
  
  def locust_init(environment, **kwargs):
      pass
  init.add_listener(locust_init)
  ```
  ```python
  # This works as well
  from locust.events import init
  
  @init.add_listener
  def _(environment, **kwargs):
      pass
  ```
- [ ] Clean up among the command line arguments.
- [ ] Rename `Locust` to `User` (see #1264 for a longer discussion)
- [ ] Remove the `Locust.setup` and `Locust.teardown` hooks in favour of adding `test_start` and `test_stop` events to `Environment.events` (#1265).
- [ ] Change how logging is handled. Currently we're wrapping sys.stdout and sys.stderr which seems wrong. 
- [ ] Remove `@seq_task` and instead add a `SequentialTaskSet` class that run all declared tasks sequentially.
- [ ] Maybe rename `host` to `base_url` (#1217)?
- [ ] Replace master/slave terminology with master/worker.
- [ ] Re-work the step load feature. I'd like it to be something more general like a run plan with an API that one can hook into. It should also be able to both spawn and kill simulated users.

These are the changes that I could think of from the top of my head right now. It's just a start and we can edit this post to keep the above list up to date.

### CODE REVIEW

**Comment by cyberw** [comment]
Cool stuff! I’ll have a deeper look soon.

**Comment by cyberw** [comment]
btw, I like all the suggested "more changes" as well (though I think most of them of deserve their own PRs instead of having a mega-PR :)

**Comment by heyman** [comment]
> btw, I like all the suggested "more changes" (though I think most of them of deserve their own PRs instead of having a mega-PR :)

You're right. We should probably branch of master in a 0.X branch, which we can use for bug fixes in the 0.14 version until we're ready to release a 1.0. 

Alternatively, we could make separate pull requests to *this* PR's branch (at least I think that's possible on Github?), though it's probably better to keep the 1.0 changes in master, and make a separate branch for the 0.14 version.

**Comment by cyberw** [comment]
> You're right. We should probably branch of master in a 0.X branch, which we can use for bug fixes in the 0.14 version until we're ready to release a 1.0.
> 
> Alternatively, we could make separate pull requests to this PR's branch (at least I think that's possible on Github?), though it's probably better to keep the 1.0 changes in master, and make a separate branch for the 0.14 version.

We can branch a emergency fix branch for 0.14 if and only if we need it. I think GH will have an easier time keeping track of PRs & stuff if everything is merged to master instead of having PRs to this branch. But either way is fine.

**Comment by heyman** [comment]
> We can branch a emergency fix branch for 0.14 if and only if we need it

Sounds good!

**Comment by heyman** [comment]
Below is a working example where Locust is used as a lib instead of started through the default command (main.py). 

Hopefully it's easier to grasp the changes to the internal API by looking at this example, compared to looking at all the changes in the PR.

```python
import gevent
from locust import HttpLocust, TaskSet, task, between
from locust.runners import LocalLocustRunner
from locust.env import Environment
from locust.stats import stats_printer
from locust.log import setup_logging
from locust.web import WebUI

setup_logging("INFO", None)

class User(HttpLocust):
    wait_time = between(1, 3)
    host = "https://docs.locust.io"
    
    class task_set(TaskSet):
        @task
        def my_task(self):
            self.client.get("/")
        
        @task
        def task_404(self):
            self.client.get("/non-existing-path")

# setup Environment and Runner
env = Environment(locust_classes=[User])
runner = LocalLocustRunner(environment=env)
# start a WebUI instance
web_ui = WebUI(runner=runner, environment=env)
gevent.spawn(lambda: web_ui.start("127.0.0.1", 8089))
# start a greenlet that periodically outputs the current stats
gevent.spawn(stats_printer(env.stats))

# start the test
runner.start(1, hatch_rate=10)

# wait for the greenlets (indefinitely)
runner.greenlet.join()
```

**Comment by cyberw** [comment]
> Below is a working example where Locust is used as a lib instead of started through the default command (main.py).

Cool stuff! Lets add a unit test that does something like that!

**Comment by cyberw** [comment]
I’m not a really huge fan of the init thing, it adds some complexity for the user for such a simple thing. Not sure if there is an alternate solution, but I’d like us to explore it. It would be best if there was an obvious ”it is just python” approach for the user

**Comment by heyman** [comment]
> I’m not a really huge fan of the init thing, it adds some complexity for the user for such a simple thing. Not sure if there is an alternate solution, but I’d like us to explore it.

I don't know if we can make it much simpler (than a single import, a single function call, and a function definition), but I'm definitely open to other suggestions. 

What I like about this solution is that it's using the event system already built into Locust. 

I'm also thinking that we could have another event (that'll get the ArgumentParser instance as argument) that end-users and plugins can hook into, that can be used to add extra command line options. Something like `locust.events.init_command_line_parser`.

**Comment by heyman** [comment]
I've now updated the documentation as well, so as soon as we've decided on the initialization API when extending Locust, I think we can merge this PR (and branch off a separate 0.14 branch), and create separate issues for the stuff under "More changes" in the OP.

**Comment by cyberw** [comment]
> > I’m not a really huge fan of the init thing, it adds some complexity for the user for such a simple thing. Not sure if there is an alternate solution, but I’d like us to explore it.
> 
> I don't know if we can make it much simpler (than a single import, a single function call, and a function definition), but I'm definitely open to other suggestions.

Isnt it  just replacing what could previously be done by placing code in the top level? that is a lot less complex than the three step-rocket you just described. Or am I missing something?

> I'm also thinking that we could have another event (that'll get the ArgumentParser instance as argument) that end-users and plugins can hook into, that can be used to add extra command line options. Something like `locust.events.init_command_line_parser`.

Cool idea! I think most use cases are better of just using something like env vars (which we should probably document somewhere), but it could be useful... maybe...

**Comment by cyberw** [comment]
I think we should try to resolve resolve the "double coupling" between locust and environment before merging. (`l = MyLocust(Environment(locust_classes=[MyLocust])` is just too strange)

**Comment by heyman** [comment]
> Isnt it just replacing what could previously be done by placing code in the top level? that is a lot less complex than the three step-rocket you just described. Or am I missing something

The reason why you could previously import events and register listeners directly at the module level was because the events were global instances. A large part of the whole refactoring has been to get rid of the global state. 

I can think of some very hacky work-arounds like using global dummy `EventHook` instances under `locust.event` that can be used to register and temporarily hold listeners *before* the `Environment` instance has been initialized, which will then be *copied* to the real `Environment.events` once the Environment has been created. However, the magic/obfuscation this would add to the code feels like a much bigger drawback than the extra code one needs to write when *extending* locust with new funcitonality (keep in mind that most end-users will never need to use the `init` event).

> I think most use cases are better of just using something like env vars to feed parameters into the testplan (which we should probably document somewhere), but it could be useful... maybe...

In many cases, absolutely. For example, I'd definitely prefer using env variables for some project specific settings. However, for something more general - let's say a a plugin for pushing stats to grafana or statsd - it'd be pretty neat to be able to inject a `--grafana-host` command line option.

**Comment by cyberw** [comment]
> > Isnt it just replacing what could previously be done by placing code in the top level? that is a lot less complex than the three step-rocket you just described. Or am I missing something
> 
> The reason why you could previously import events and register listeners directly at the module level was because the events were global instances. A large part of the whole refactoring has been to get rid of the global state.
> 
> I can think of some very hacky work-arounds like using global dummy `EventHook` instances under `locust.event` that can be used to register and temporarily hold listeners _before_ the `Environment` instance has been initialized, which will then be _copied_ to the real `Environment.events` once the Environment has been created. However, the magic/obfuscation this would add to the code feels like a much bigger drawback than the extra code one needs to write when _extending_ locust with new funcitonality (keep in mind that most end-users will never need to use the `init` event).
> 

I do like the new method in general, I was just thinking about the init event (which I think wasnt really needed before, right?). either way, it is not a showstopper for me.

> > I think most use cases are better of just using something like env vars to feed parameters into the testplan (which we should probably document somewhere), but it could be useful... maybe...
> 
> In many cases, absolutely. For example, I'd definitely prefer using env variables for some project specific settings. However, for something more general - let's say a a plugin for pushing stats to grafana or statsd - it'd be pretty neat to be able to inject a `--grafana-host` command line option.

(y)

**Comment by heyman** [comment]
> I think we should try to resolve resolve the "double coupling" between locust and environment before merging. (`l = MyLocust(Environment(locust_classes=[MyLocust]`) is just too strange/smelly)

I've definitely thought quite a bit about that :). I'd be very happy to hear suggestions. The reason for it is because the locust user instances needs a reference to the `Environment` in order to be able to trigger `request_success and `request_failure` (which is available in `Environment.events`
) as well as retrieve the `host`/`base_url`. 

To be fair you wouldn't really see the above code anywhere outside  the unit tests, since it's the runner that instantiates the Locust users (and then passes in a reference to the environment instance).

**Comment by cyberw** [comment]
> > I think we should try to resolve resolve the "double coupling" between locust and environment before merging. (`l = MyLocust(Environment(locust_classes=[MyLocust]`) is just too strange/smelly)
> 
> I've definitely thought quite a bit about that :). I'd be very happy to hear suggestions. The reason for it is because the locust user instances needs a reference to the `Environment` in order to be able to trigger `request_success and `request_failure`(which is available in`Environment.events`) as well as retrieve the`host`/`base_url`.
> 
> To be fair you wouldn't really see the above code anywhere outside the unit tests, since it's the runner that instantiates the Locust users (and then passes in a reference to the environment instance).

I can have a look tomorrow, but I think it would need some decent sized changes so I think you are better equipped as you wrote it :)

It is not reading the particular line of code that is a problem, the problem is that I think Environment is so tightly linked to Locust that they are essentially behaving as if they were the same class (which will harm maintainability)

**Comment by heyman** [comment]
> I do like the new method in general, I was just thinking about the init event (which I think wasnt really needed before, right?).

Yep, the reason it wasn't needed before was because of the global event variables. This PR moves the events onto the Environment instance which is instantiated *after* the end-user's locustfile is imported. So at the time the locustfile is evaluated the EventHook instances doesn't exist.

One alternative could be to instantiate the Events class (and thereby the EventHook classes) separately in `main.py`, and store that instance is some global variable accessible to the locustfile, *before* we create the `Environment`. Once the `Environment` is created we can set it's `events` attribute to the already created `Events` instance. That would allow you to still register listeners at the module level:

```python
# locustfile.py
from locust.events import events

@events.request_success.add_listener
def my_listener(**kw):
    pass
```

Hm, maybe that would be better?

It would make the API more different compared to when using Locust as a lib (then one would access the events through `environment.events`)

**Comment by heyman** [comment]
> It is not reading the particular line of code that is a problem, the problem is that I think Environment is so tightly linked to Locust that they are essentially behaving as if they were the same class (which will harm maintainability)

I've played with the idea of just injecting the events and the host/base_url into the Locust instances (instead of a reference to the whole Environment). However I keep thinking that it can be useful to have access to more info about the environment that a Locust user is being run in. Also, the plain `Locust` class (as compared to `HttpLocust`) doesn't even need the `host` variable.

**Comment by cyberw** [comment]
Maybe Environment and LocustRunner are really the same class? Most methods in LocustRunner operate on its .environment and .stats fields (stats is also available on environment, another strange situation)

Not sure if it would help with the issue (Locust needing an Environment to be created, which in turn needs a list of locust classes), but it would cut down on the complexity a little. I think the names "Environment" and "LocustRunner" are a bit vague, perhaps choosing better names will help us divide the responsibility between them.

**Comment by heyman** [comment]
> Maybe Environment and LocustRunner are really the same class?

Every attribute of the Environment could be moved onto the Runner class, but I don't think that would result in a better API or a more maintainable code base. 

I see the Environment class as a holder of the configuration for the runner (as well as the Locust classes). The Runner is dependent on the Environment instance, but not the other way around (even though the environment does hold a reference to the Runner which is using it).

**Comment by heyman** [comment]
After some discussion with @cyberw on Slack I've now reworked the API slightly.

* `Environemnt.locust_classes` has been moved to `LocustRunner.locust_classes`, and `LocustRunner` now takes `locust_classes` as a constructor argument in addition to an `Environment` instance.
* `Environment.stats` has been moved to `LocustRunner.stats`.
* `Environment` does no longer hold references to the `LocustRunner` instance or the `WebUI` instance (avoids circular class dependencies). Locust scripts and plugins that need access to these instances will instead have to retrieve them by listening to an `init` event which will be passed the instances as event arguments.

**Comment by cyberw** [comment]
Maybe we should rename this PR and try to merge before it grows too big? :)

**Comment by heyman** [comment]
Yes, that sounds like a good idea. Though we should branch off the master into a 0.x branch (that we can use for critical bug fixes) before merging this.

**Comment by heyman** [comment]
This has now been merged to master. I've created a new branch (v0.x) from the previous master.

I've created a 1.0 milestone, and created stub issues for some of the items in my checkbox list under "More changed" in the OP.

File: examples/custom_xmlrpc_client/xmlrpc_locustfile.py [hunk 1]
```diff
@@ -1,27 +1,30 @@
...
 
-class XmlRpcClient(xmlrpclib.ServerProxy):
+class XmlRpcClient(ServerProxy):
     """
     Simple, sample XML RPC client implementation that wraps xmlrpclib.ServerProxy and 
     fires locust events on request_success and request_failure, so that all requests 
     gets tracked in locust's statistics.
     """
+    
+    _locust_environment = None
```
[the remark this answers is not in the document]
Comment (heyman) [reply examples/custom_xmlrpc_client/xmlrpc_locustfile.py:14]: I just wanted to make sure that we wouldn't clash with anything in `xmlrpclib.ServerProxy`.

File: locust/clients.py [hunk 2]
```diff
@@ -186,9 +187,10 @@ class ResponseContextManager(LocustResponse):
     
     _is_reported = False
     
-    def __init__(self, response):
+    def __init__(self, response, environment):
         # copy data from response to this object
         self.__dict__ = response.__dict__
+        self.locust_environment = environment
```
[the remark this answers is not in the document]
Comment (heyman) [reply locust/clients.py]: I prefixed it with `locust_` just to make sure that we won't clash with any (current or future) attribute in requests.Session.

File: locust/test/test_locust_class.py [hunk 3]
```diff
@@ -256,8 +258,8 @@ class MyLocust2(Locust):
             host = "http://127.0.0.1"
             task_set = MyTaskSet2
         
-        l = MyLocust()
-        l2 = MyLocust2()
+        l = MyLocust(Environment(locust_classes=[MyLocust]))
```
[the remark this answers is not in the document]
Comment (heyman) [reply locust/test/test_locust_class.py]: Hm, maybe. Not sure though. One would never instantiate a Locust class like that except for in a test.

[the remark this answers is not in the document]
Comment (cyberw) [reply locust/test/test_locust_class.py]: I think this is one of those things we spend a lot of time discussing later, so it is kind of old news :)

File: locust/clients.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cyberw) [reply locust/clients.py]: yea, this is gone now anyways :)

**Pull request merged**